### ISSUE TITLE
Running example in parallel 

### ISSUE THREAD

**Issue by bonh** [ibaned/omega_h-dolfin#1]
I am really interested in and eager to help getting your work up and running in conjunction with dolfin. 

docker:
`docker run -ti -v $(pwd):/home/fenics/shared -w /home/fenics/shared quay.io/dolfinadjoint/dolfin-adjoint:latest`

Inside the docker I clone your repos `omega_h` and `omega_h-dolfin`, and execute the build scripts in `omega_h-dolfin/aux`.

On one core, it seems to work:
```
Solving linear variational problem.
limited gradation in 10 steps
generated metrics in 0.0124711 seconds
before adapting:
2048 triangles, quality [0.43,1.00], 2048 >0.40
3136 edges, length [0.11,1.85], 2492 <0.71, 634 in [0.71,1.41], 10 >1.41
quality histogram:
0.00-0.10: 0
0.10-0.20: 0
0.20-0.30: 0
0.30-0.40: 0
0.40-0.50: 5
0.50-0.60: 21
0.60-0.70: 50
0.70-0.80: 95
0.80-0.90: 1764
0.90-1.00: 113
length histogram:
0.00-0.30: 601
0.30-0.60: 1428
0.60-0.90: 902
0.90-1.20: 184
1.20-1.50: 17
1.50-1.80: 3
1.80-2.10: 1
2.10-2.40: 0
2.40-2.70: 0
2.70-3.00: 0
addressing edge lengths
refining 9 edges
2066 triangles, quality [0.43,1.00], 2066 >0.40
3163 edges, length [0.11,1.44], 2500 <0.71, 662 in [0.71,1.41], 1 >1.41
coarsening 238 vertices
1617 triangles, quality [0.31,0.99], 12 in [0.30,0.40], 1605 >0.40
2476 edges, length [0.11,1.44], 1626 <0.71, 849 in [0.71,1.41], 1 >1.41
refining 1 edges
1618 triangles, quality [0.31,0.99], 12 in [0.30,0.40], 1606 >0.40
2478 edges, length [0.11,1.41], 1627 <0.71, 851 in [0.71,1.41]
coarsening 209 vertices
1202 triangles, quality [0.32,1.00], 5 in [0.30,0.40], 1197 >0.40
1853 edges, length [0.11,1.41], 970 <0.71, 883 in [0.71,1.41]
coarsening 118 vertices
970 triangles, quality [0.32,1.00], 2 in [0.30,0.40], 968 >0.40
1503 edges, length [0.11,1.41], 615 <0.71, 888 in [0.71,1.41]
coarsening 82 vertices
808 triangles, quality [0.32,1.00], 5 in [0.30,0.40], 803 >0.40
1259 edges, length [0.16,1.41], 376 <0.71, 883 in [0.71,1.41]
coarsening 42 vertices
725 triangles, quality [0.34,1.00], 4 in [0.30,0.40], 721 >0.40
1134 edges, length [0.16,1.41], 237 <0.71, 897 in [0.71,1.41]
coarsening 26 vertices
674 triangles, quality [0.39,1.00], 2 in [0.30,0.40], 672 >0.40
1057 edges, length [0.17,1.41], 162 <0.71, 895 in [0.71,1.41]
coarsening 11 vertices
652 triangles, quality [0.37,1.00], 2 in [0.30,0.40], 650 >0.40
1024 edges, length [0.22,1.41], 132 <0.71, 892 in [0.71,1.41]
coarsening 7 vertices
638 triangles, quality [0.37,1.00], 1 in [0.30,0.40], 637 >0.40
1003 edges, length [0.25,1.41], 116 <0.71, 887 in [0.71,1.41]
coarsening 1 vertices
636 triangles, quality [0.37,1.00], 1 in [0.30,0.40], 635 >0.40
1000 edges, length [0.40,1.41], 112 <0.71, 888 in [0.71,1.41]
coarsening 1 vertices
634 triangles, quality [0.37,1.00], 1 in [0.30,0.40], 633 >0.40
997 edges, length [0.40,1.41], 110 <0.71, 887 in [0.71,1.41]
addressing element qualities
swapping 2 2D edges
634 triangles, quality [0.43,1.00], 634 >0.40
997 edges, length [0.40,1.41], 110 <0.71, 887 in [0.71,1.41]
quality histogram:
0.00-0.10: 0
0.10-0.20: 0
0.20-0.30: 0
0.30-0.40: 0
0.40-0.50: 3
0.50-0.60: 9
0.60-0.70: 40
0.70-0.80: 49
0.80-0.90: 401
0.90-1.00: 132
length histogram:
0.00-0.30: 0
0.30-0.60: 17
0.60-0.90: 399
0.90-1.20: 364
1.20-1.50: 217
1.50-1.80: 0
1.80-2.10: 0
2.10-2.40: 0
2.40-2.70: 0
2.70-3.00: 0
addressing edge lengths took 0.0365463 seconds
addressing element qualities took 0.00061745 seconds
adapting took 0.0378672 seconds

Solving linear variational problem.
limited gradation in 2 steps
generated metrics in 0.00164517 seconds
before adapting:
634 triangles, quality [0.43,1.00], 634 >0.40
997 edges, length [0.44,1.98], 116 <0.71, 861 in [0.71,1.41], 20 >1.41
quality histogram:
0.00-0.10: 0
0.10-0.20: 0
0.20-0.30: 0
0.30-0.40: 0
0.40-0.50: 5
0.50-0.60: 7
0.60-0.70: 36
0.70-0.80: 50
0.80-0.90: 386
0.90-1.00: 150
length histogram:
0.00-0.30: 0
0.30-0.60: 18
0.60-0.90: 406
0.90-1.20: 355
1.20-1.50: 212
1.50-1.80: 4
1.80-2.10: 2
2.10-2.40: 0
2.40-2.70: 0
2.70-3.00: 0
addressing edge lengths
refining 15 edges
660 triangles, quality [0.43,1.00], 660 >0.40
1038 edges, length [0.35,1.51], 133 <0.71, 900 in [0.71,1.41], 5 >1.41
coarsening 13 vertices
634 triangles, quality [0.43,1.00], 634 >0.40
999 edges, length [0.44,1.51], 108 <0.71, 886 in [0.71,1.41], 5 >1.41
refining 5 edges
642 triangles, quality [0.43,1.00], 642 >0.40
1012 edges, length [0.35,1.41], 118 <0.71, 894 in [0.71,1.41]
coarsening 3 vertices
636 triangles, quality [0.43,1.00], 636 >0.40
1003 edges, length [0.39,1.41], 111 <0.71, 892 in [0.71,1.41]
quality histogram:
0.00-0.10: 0
0.10-0.20: 0
0.20-0.30: 0
0.30-0.40: 0
0.40-0.50: 6
0.50-0.60: 9
0.60-0.70: 35
0.70-0.80: 55
0.80-0.90: 372
0.90-1.00: 159
length histogram:
0.00-0.30: 0
0.30-0.60: 21
0.60-0.90: 402
0.90-1.20: 368
1.20-1.50: 212
1.50-1.80: 0
1.80-2.10: 0
2.10-2.40: 0
2.40-2.70: 0
2.70-3.00: 0
addressing edge lengths took 0.00442859 seconds
addressing element qualities took 2.834e-06 seconds
adapting took 0.00492576 seconds

Solving linear variational problem.
Attempting to use an MPI routine after finalizing MPICH
```
Although, the last line seems odd...

But on two cores:
```
assertion mesh->owners_have_all_upward(VERT) failed at /home/fenics/omega_h/src/Omega_h_recover.cpp +152
assertion mesh->owners_have_all_upward(VERT) failed at /home/fenics/omega_h/src/Omega_h_recover.cpp +152
```
Do you have an idea what is happening?

**Comment by bonh** [comment]
Oh, just noticed, that `mpirun -np 2 ./mox_test` from [omega_h](https://github.com/ibaned/omega_h) `omega_h/src` fails too with 
```
using tau = 10
assertion mesh->have_all_upward() failed at /home/fenics/shared/omega_h_suite/omega_h/src/Omega_h_metric.cpp +670
```

**Comment by ibaned** [comment]
@bonh thanks for posting this issue with detailed information! It was especially helpful to have the Docker command:
```
docker run -ti -v $(pwd):/home/fenics/shared -w /home/fenics/shared quay.io/dolfinadjoint/dolfin-adjoint:latest
```
That saved me a lot of time in debugging this. I have pushed a new Omega_h version `v9.9.0` and made improvements to the `omega_h-dolfin` repository. These fix most of the issue, now both examples run. The current fixes have to do with the fact that Omega_h sometimes needs ghost elements and sometimes not, and users have to manually choose when to create ghost elements.

However, an issue remains that the simulation doesn't give the right answer in parallel (see the attached figure showing the imprint of the partioning on the solution). I believe this is because I don't properly copy all the sharing information to DOLFIN. While trying to fix this, I ran into a serious obstable: DOLFIN requires all non-shared entities to show up first in the local order, followed by all shared entities. Omega_h doesn't operate this way, so I will need to do more work to get things to fully work in parallel.
![screen shot 2018-02-02 at 1 44 00 pm](https://user-images.githubusercontent.com/971921/35755990-bb39172c-0826-11e8-81cb-fef3054c20e6.png)

I will post again here as soon as I have a full fix in place.

**Comment by bonh** [comment]
I can confirm your observation - nice work! If I can do anything to help please let me know.

**Comment by ibaned** [comment]
I made more progress over the weekend, and got my local ordering to match what I think DOLFIN wants, but the parallel output still looks wrong. I think my next step is to use DOLFIN's functions for partitioning a mesh, and observe what the resulting data structure looks like.

**Comment by ibaned** [comment]
Okay, I finally figured out the missing piece. DOLFIN needs to be told which facets are connected to two cells in a global sense even if locally they are only connected to one. I have just released version `9.9.1` of Omega_h which should work correctly in parallel.

However, note that my current fix will not work quite right if there are degrees of freedom on edges or faces that Omega_h needs to remap. Since Omega_h isn't able to remap such fields currently anyway this isn't a big concern, and the current approach should support many common use cases.

@bonh if you confirm that this fix is sufficient, we can close this issue. Thanks again for reporting.
